include/ignition/rendering/base/BaseGeometry.hh

```diff
@@ -89,12 +97,37 @@ namespace ignition
       if (material) this->SetMaterial(material, _unique);
     }
 
+    //////////////////////////////////////////////////
+    template <class T>
+    GeometryPtr BaseGeometry<T>::Clone() const
+    {
+      if (!this->Scene())
+      {
+        ignerr << "Cloning a geometry failed because the geometry to be "
+          << "cloned does not belong to a scene.\n";
+        return nullptr;
+      }
+      else if (this->meshName.empty())
+      {
+        ignerr << "Cloning a geometry failed because the name of the mesh is "
+          << "missing.\n";
+        return nullptr;
+      }
+
+      auto result = this->Scene()->CreateMesh(this->meshName);
+      if (result && this->Material())
+        result->SetMaterial(this->Material());
```

Comment:
I think we should. Users could be setting the submesh material directly. So while the mesh's material is empty, it's possible that the submesh may have a material set.

Reply:
After thinking about how to handle submeshes, I believe that there may be a design issue in this PR. In order to handle submeshes, I would have to cast `this` from a `Geometry` to a `Mesh`, since a `Mesh` inherits a `Geometry`. So, this makes me feel like the code for cloning meshes should not be in the `Geometry` class, since a mesh is a geometry, but a geometry is not a mesh (I don't think I've ever seen code where a parent class tries to call code that's unique to the child class). Another reason why I think the design/architecture here may be incorrect is because I am making the assumption that the geometry has a mesh, but I don't think I should assume this because the geometry isn't required to have a mesh. So, I think that the design/API structure for cloning geometries needs to be modified in order to handle submeshes. Does this make sense? Do you have any thoughts about how to solve this issue?

Reply:
I see. We can have a `Geometry::Clone()` and a  `BaseGeometry::Clone` function does not do anything by default and let the derived classes override this function, e.g.Mesh::Clone() / BaseMesh::Clone()

Reply:
I went ahead and followed your suggestion by placing the clone implementation in the `Mesh` class: https://github.com/ignitionrobotics/ign-rendering/pull/397/commits/f5b07ea19707d9f2ba97565f5d7d5140d496f238

However, I think that this still isn't complete. For example, when I try to copy something like an ellipse or capsule in the GUI (I'm testing this with https://github.com/ignitionrobotics/ign-gazebo/pull/1013), I see the following error: 
```
[GUI] [Err] [Ogre2Visual.cc:130] Cannot attach null geometry.
```

I believe that's because certain geometries aren't a `Mesh`, which means that the `Geometry::Clone` implementation gets called, which does nothing (it returns `nullptr`): https://github.com/ignitionrobotics/ign-rendering/blob/f5b07ea19707d9f2ba97565f5d7d5140d496f238/include/ignition/rendering/base/BaseGeometry.hh#L101

So, do I need to implement the `Clone` method for other geometry child classes as well? I believe that I should probably do this for the [`Capsule` class](https://github.com/ignitionrobotics/ign-rendering/blob/main/include/ignition/rendering/Capsule.hh), but are there any other classes I'm missing?

Reply:
As a follow-up to my previous comment, I tried to implement cloning functionality for capsule geometries: https://github.com/ignitionrobotics/ign-rendering/pull/397/commits/9d87388a15869b289e73ec66f3f9544a09228bd1

However, I am seeing compile-time warnings/errors about using an incomplete `Scene` type. I'm not sure why this is happening for the capsule, because it was fine with the mesh class. Perhaps the inheritance works differently for capsules? I tried to figure out how capsule inheritance structure differs from the mesh class, but couldn't seem to figure it out.

Here are the compile-time warnings:

<details>
  <summary>Compilation warnings</summary>
  
```
--- stderr: ignition-rendering6                                  
In file included from /ws/src/ign-rendering/ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:24,
                 from /ws/src/ign-rendering/ogre2/src/Ogre2Marker.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh: In member function 'virtual ignition::rendering::v6::GeometryPtr ignition::rendering::v6::BaseCapsule<T>::Clone() const':
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:123:34: warning: invalid use of incomplete type 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
       auto result = this->Scene()->CreateCapsule();
                                  ^~
In file included from /ws/src/ign-rendering/include/ignition/rendering/Geometry.hh:23,
                 from /ws/src/ign-rendering/include/ignition/rendering/Capsule.hh:22,
                 from /ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:23,
                 from /ws/src/ign-rendering/ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:24,
                 from /ws/src/ign-rendering/ogre2/src/Ogre2Marker.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/RenderTypes.hh:84:11: note: forward declaration of 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
     class Scene;
           ^~~~~
In file included from /ws/src/ign-rendering/ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:24,
                 from /ws/src/ign-rendering/ogre2/src/Ogre2Capsule.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh: In member function 'virtual ignition::rendering::v6::GeometryPtr ignition::rendering::v6::BaseCapsule<T>::Clone() const':
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:123:34: warning: invalid use of incomplete type 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
       auto result = this->Scene()->CreateCapsule();
                                  ^~
In file included from /ws/src/ign-rendering/include/ignition/rendering/Geometry.hh:23,
                 from /ws/src/ign-rendering/include/ignition/rendering/Capsule.hh:22,
                 from /ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:23,
                 from /ws/src/ign-rendering/ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:24,
                 from /ws/src/ign-rendering/ogre2/src/Ogre2Capsule.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/RenderTypes.hh:84:11: note: forward declaration of 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
     class Scene;
           ^~~~~
In file included from /ws/src/ign-rendering/ogre/include/ignition/rendering/ogre/OgreCapsule.hh:22,
                 from /ws/src/ign-rendering/ogre/src/OgreMarker.cc:20:
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh: In member function 'virtual ignition::rendering::v6::GeometryPtr ignition::rendering::v6::BaseCapsule<T>::Clone() const':
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:123:34: warning: invalid use of incomplete type 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
       auto result = this->Scene()->CreateCapsule();
                                  ^~
In file included from /ws/src/ign-rendering/include/ignition/rendering/Geometry.hh:23,
                 from /ws/src/ign-rendering/include/ignition/rendering/Capsule.hh:22,
                 from /ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:23,
                 from /ws/src/ign-rendering/ogre/include/ignition/rendering/ogre/OgreCapsule.hh:22,
                 from /ws/src/ign-rendering/ogre/src/OgreMarker.cc:20:
/ws/src/ign-rendering/include/ignition/rendering/RenderTypes.hh:84:11: note: forward declaration of 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
     class Scene;
           ^~~~~
In file included from /ws/src/ign-rendering/ogre/include/ignition/rendering/ogre/OgreCapsule.hh:22,
                 from /ws/src/ign-rendering/ogre/src/OgreCapsule.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh: In member function 'virtual ignition::rendering::v6::GeometryPtr ignition::rendering::v6::BaseCapsule<T>::Clone() const':
/ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:123:34: warning: invalid use of incomplete type 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
       auto result = this->Scene()->CreateCapsule();
                                  ^~
In file included from /ws/src/ign-rendering/include/ignition/rendering/Geometry.hh:23,
                 from /ws/src/ign-rendering/include/ignition/rendering/Capsule.hh:22,
                 from /ws/src/ign-rendering/include/ignition/rendering/base/BaseCapsule.hh:23,
                 from /ws/src/ign-rendering/ogre/include/ignition/rendering/ogre/OgreCapsule.hh:22,
                 from /ws/src/ign-rendering/ogre/src/OgreCapsule.cc:23:
/ws/src/ign-rendering/include/ignition/rendering/RenderTypes.hh:84:11: note: forward declaration of 'using element_type = class ignition::rendering::v6::Scene' {aka 'class ignition::rendering::v6::Scene'}
     class Scene;
           ^~~~~

```
</details>